src/peer-store/key-book.js

```diff
@@ -0,0 +1,83 @@
+'use strict'
+
+const errcode = require('err-code')
+const debug = require('debug')
+const log = debug('libp2p:peer-store:key-book')
+log.error = debug('libp2p:peer-store:key-book:error')
+
+const PeerId = require('peer-id')
+
+const Book = require('./book')
+
+const {
+  codes: { ERR_INVALID_PARAMETERS }
+} = require('../errors')
+
+/**
+ * The KeyBook is responsible for keeping the known public keys of a peer.
+ */
+class KeyBook extends Book {
+  /**
+  * @constructor
+  * @param {PeerStore} peerStore
+  */
+  constructor (peerStore) {
+    super({
+      peerStore,
+      eventName: 'change:pubkey', // TODO: the name is not probably the best!?
+      eventProperty: 'pubkey',
+      eventTransformer: (data) => data.pubKey
+    })
+
+    /**
+     * Map known peers to their known Public Key.
+     * @type {Map<string, PeerId>}
+     */
+    this.data = new Map()
+  }
+
+  /**
+   * Set PeerId. If the peer was not known before, it will be added.
+   * @override
+   * @param {PeerId} peerId
+   * @return {KeyBook}
+ */
+  set (peerId) {
+    if (!PeerId.isPeerId(peerId)) {
+      log.error('peerId must be an instance of peer-id to store data')
+      throw errcode(new Error('peerId must be an instance of peer-id'), ERR_INVALID_PARAMETERS)
+    }
+
+    const id = peerId.toB58String()
+    const recPeerId = this.data.get(id)
+
+    !recPeerId && this._ps.emit('peer', peerId)
+    // If no record available, or it is incomplete
+    if (!recPeerId || (peerId.pubKey && !recPeerId.pubKey)) {
+      this._setData(peerId, peerId, {
+        emit: Boolean(peerId.pubKey) // No persistence if no public key
+      })
+      log(`stored provided public key for ${id}`)
+    }
```

Comment:
This block is a bit confusing. I think it's trying to cover too many use cases?
1. we have no record of the peer
1.1 we have no record of the peer, and we dont have a public key to add so dont emit (why are we calling set on the keybook if this is the case?)
2. we have an existing record and a new public key but no existing public key (why do we have a record with no public key in the keystore?)

I think the keystore is also trying to keep track of peer ids, we should just avoid this. Let the peer store do that and just make it a computed property across the books, because if we have no addresses/protocols/keys/etc associated to a peer id, it really shouldn't exist for us as it has 0 value.